copy _UnitRegistry if possible instead of completly rebuilding it.

MSeifert04 opened the pull request:
This is somewhat inspired by @mhvk (https://github.com/astropy/astropy/pull/5304#discussion_r77541744).

This PR significantly speeds up the units-core-module functions: `add_enabled_units`, `set_enabled_equivalencies` and `add_enabled_equivalencies`.

The change I made here is to assume that when a new `_UnitRegistry` is built from another `_UnitRegistry` we don't need to go through the `add_enabled_units` method (which takes more than 50ms on my computer) but can simply copy the values from the old registry.

Some test cases:

```
from astropy.analytic_functions import blackbody_nu
from astropy import units as u
import numpy as np

wave = np.logspace(5, 8, 100000) * u.AA
temp = 100. * u.K

%timeit blackbody_nu(wave, temp)
# 10 loops, best of 3: 115 ms per loop   (before)
# 10 loops, best of 3: 48.6 ms per loop  (after)


def test():
    # This is one part of the "blackbody_nu" function
    with u.add_enabled_equivalencies(u.spectral() + u.temperature()):
        freq = u.Quantity(wave, u.Hz, dtype=np.float64)
        temperature = u.Quantity(temp, u.K, dtype=np.float64)

%timeit test()
# 10 loops, best of 3: 72.9 ms per loop   (before)
# 100 loops, best of 3: 7.97 ms per loop  (after)


def test2():
    with u.imperial.enable():  # uses units.core.add_enabled_units
        pass

%timeit test2()
# 10 loops, best of 3: 68.1 ms per loop   (before)
# 100 loops, best of 3: 3.45 ms per loop  (after)
```

mhvk commented on astropy/units/core.py:
```diff
@@ -113,14 +114,25 @@ class _UnitRegistry(object):
-            equivalencies = init.equivalencies
-            init = init.all_units
-
         self._reset_units()
```
These two lines can also be moved to the `else` clause.

Which means the two branches are completely separate, which in turn makes me wonder if it wouldn't be cleaner to implement a `copy` method and use that instead. But maybe not quite worth the effort (and can be done separately anyway).

MSeifert04 replied:
I thought about a `copy` method but decided against it because it's a private class (leading underscore) and only `add/set_enabled_units/equivalencies` do use it.

I got sidetracked by trying to implement it as `ChainMap` (which would also make all the copies unnecessary) until I realized that it's not avaiable in python2. :smile: Need to think about it some more. :smile:

mhvk replied:
That would be brilliant! It seems one can get a backport via https://pypi.python.org/pypi/chainmap/1.0.0 (oddly, it seems to suggest it should exist on python 2.7...

And at least on my machine, I seem to be able to do:
```
from distlib.compat import ChainMap
```
which comes from `python-distlib` (which I think astropy relies on, but perhaps only for packaging).

mhvk commented:
This is great! I hope there are more low-hanging (though this was by no means obvious...) fruit to be found!

There is the one comment about moving the two lines. Don't worry about a `copy` method -- the more I think about it, the less I'm convinced this is called for. If anything, I think it is worth considering whether one could copy only when it is actually needed. I.e., one would just set the items to equal, and only in  `add_enabled_equivalencies` or `add_enabled_units` make the actual copy. But maybe that is better for another iteration, and might be even better done by separating out units and equivalencies (since it is almost invariably only one of those that gets added to).

However, do add an entry about the speed-up in the `Other Changes and Additions` of `CHANGES.rst` -- this is well worth it!

pllim commented:
I wonder if this will also solve #4826...

MSeifert04 commented:
@pllim Want to test? I don't have great hopes because these functions are rarely used in tests/core. If it weren't for the `blackbody_nu` I probably wouldn't have found the bottleneck so easily.

@mhvk I've added a changelog and moved the two `reset`-lines. I hoped this would make it into 1.2.2 since it doesn't contain API changes. But if it's scheduled for 1.3 I can give it a try with `ChainMap`. 

Some comments about the seperation of units and equivalencies: 

- The `set_enabled_units` doesn't copy the `units`-related dictionaries because the function only passes the `equivalencies`.

- The `set_enabled_equivalencies` copies the existing equivalencies but by default there is no equivalency (so negligable overhead with copy): Profiled for `with u.add_enabled_equivalencies(u.spectral() + u.temperature()):`

```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   122         1           14     14.0      0.5              self._equivalencies = init._equivalencies.copy()
   123         1          590    590.0     21.5              self._all_units = init._all_units.copy()
   124         1         1591   1591.0     58.1              self._registry = init._registry.copy()
   125         1           41     41.0      1.5              self._non_prefix_units = init._non_prefix_units.copy()
   126                                                       # The physical type is a dictionary containing sets as values.
   127                                                       # All of these must be copied otherwise we could alter the old
   128                                                       # registry.
   129         1           10     10.0      0.4              self._by_physical_type = {k: v.copy() for k, v in
   130         1          484    484.0     17.7                                        six.iteritems(init._by_physical_type)}
```

So short of implementing it without copies it probably cannot be sped up (much) further.

mhvk commented:
I think this has to go for 1.3, since we're not solving a bug.

Since as is, this is already a big speed improvement for the most common cases (for me where an equivalency is used temporarily), I'm quite happy to just merge this when travis has passed as well. We can then consider `ChainMap` independently (I do quite like the idea...)

mhvk commented:
OK, it passed, so I am merging.  (Happily, it also passed on numpy-dev, so those PRs did the trick too.)
